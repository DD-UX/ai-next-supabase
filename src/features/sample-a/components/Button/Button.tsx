"use client";

import React from 'react';

type ButtonProps = {
    label: string;
    onClick: () => void;
    disabled?: boolean;
};

const Button = ({ label, onClick, disabled = false }: ButtonProps) => {
    return (
        <button className="bg-blue-500 text-white font-bold py-2 px-4 rounded"
                onClick={onClick}
                disabled={disabled}>
            {label}
        </button>
    );
};

export default Button;
