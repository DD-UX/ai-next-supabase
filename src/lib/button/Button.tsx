"use client";

import type { ButtonHTMLAttributes, PropsWithChildren } from "react";
import { forwardRef, useLayoutEffect, useState } from "react";
import { motion, useAnimate } from "motion/react";

import { cn } from "@/lib/utils";

type ButtonProps = PropsWithChildren<
  ButtonHTMLAttributes<HTMLButtonElement> & {
    variant?: "primary" | "secondary" | "link" | "success" | "warning" | "error";
  }
>;

// Style configuration for button variants
const baseClasses =
  "flex min-w-[120px] cursor-pointer items-center justify-center gap-2 rounded-md px-4 py-2 text-sm font-medium ring-offset-2 transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed";

const variantClasses = {
  primary: "bg-indigo-600 text-white hover:bg-indigo-700 focus-visible:ring-indigo-500",
  secondary:
    "border border-slate-300 bg-transparent text-slate-900 hover:bg-slate-100 focus-visible:ring-slate-400",
  link: "text-indigo-600 underline-offset-4 hover:underline focus-visible:ring-indigo-500",
  success: "bg-green-600 text-white hover:bg-green-700 focus-visible:ring-green-500",
  warning: "bg-amber-500 text-white hover:bg-amber-600 focus-visible:ring-amber-500",
  error: "bg-red-600 text-white hover:bg-red-700 focus-visible:ring-red-500",
};

// Disabled state styles - consistent across all variants
const disabledClasses = "disabled:bg-slate-500 disabled:text-white/80";

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ children, variant = "primary", className, disabled, ...props }, ref) => {
    const [scope, animate] = useAnimate<HTMLButtonElement>();
    const [status, setStatus] = useState<"idle" | "loading" | "success">("idle");

    useLayoutEffect(() => {
      if (typeof ref === "function") {
        ref(scope.current);
      } else if (ref && "current" in ref) {
        ref.current = scope.current;
      }
    }, [ref, scope]);

    const animateLoading = async () => {
      await animate(
        ".loader",
        { width: "20px", scale: 1, display: "block" },
        { duration: 0.2 },
      );
    };

    const animateSuccess = async (): Promise<void> => {
      await animate(".loader", { width: "0px", scale: 0, display: "none" }, { duration: 0.2 });
      await animate(
        ".check",
        { width: "20px", scale: 1, display: "block" },
        { duration: 0.2 },
      );
      await animate(
        ".check",
        { width: "0px", scale: 0, display: "none" },
        { delay: 2, duration: 0.2 },
      );
    };

    const { onClick, ...restProps } = props;

    const handleClick = async (event: React.MouseEvent<HTMLButtonElement>): Promise<void> => {
      if (status !== "idle") return;

      const onClickPromise = onClick?.(event);

      if (onClick) {
        setStatus("loading");
        await animateLoading();

        if (onClickPromise && typeof onClickPromise.then === "function") {
          // eslint-disable-next-line @typescript-eslint/await-thenable
          await onClickPromise;
        }

        setStatus("success");
        await animateSuccess();

        setTimeout(() => setStatus("idle"), 2200); // Reset after animation
      }
    };

    const isButtonDisabled = disabled || status !== "idle";

    return (
      <motion.button
        ref={scope}
        disabled={isButtonDisabled}
        className={cn(
          baseClasses,
          variantClasses[variant],
          isButtonDisabled && disabledClasses,
          className,
        )}
        {...restProps}
        // eslint-disable-next-line @typescript-eslint/no-misused-promises
        onClick={handleClick}
      >
        <motion.div layout className="flex items-center gap-2">
          <Loader />
          <CheckIcon />
          <motion.span layout>{children}</motion.span>
          <span className="sr-only" aria-live="polite" aria-atomic="true">
            {status === "loading" && "Loading, please wait."}
            {status === "success" && "Action completed successfully."}
          </span>
        </motion.div>
      </motion.button>
    );
  },
);

Button.displayName = "Button";

const Loader = (): JSX.Element => (
  <motion.svg
    animate={{ rotate: [0, 360] }}
    initial={{ scale: 0, width: 0, display: "none" }}
    transition={{ duration: 0.5, repeat: Infinity, ease: "linear" }}
    className="loader"
    width="20"
    height="20"
    viewBox="0 0 24 24"
    fill="none"
    stroke="currentColor"
    strokeWidth="2"
    strokeLinecap="round"
    strokeLinejoin="round"
  >
    <path d="M12 3a9 9 0 1 0 9 9" />
  </motion.svg>
);

const CheckIcon = (): JSX.Element => (
  <motion.svg
    initial={{ scale: 0, width: 0, display: "none" }}
    className="check"
    width="20"
    height="20"
    viewBox="0 0 24 24"
    fill="none"
    stroke="currentColor"
    strokeWidth="2"
    strokeLinecap="round"
    strokeLinejoin="round"
  >
    <path d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0" />
    <path d="M9 12l2 2l4 -4" />
  </motion.svg>
);

export default Button;
