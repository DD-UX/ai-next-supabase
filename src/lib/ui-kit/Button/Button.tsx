'use client';

import { forwardRef, type ButtonHTMLAttributes } from 'react';

type ButtonProps = ButtonHTMLAttributes<HTMLButtonElement> & {
  isLoading?: boolean;
};

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ children, isLoading, ...props }, ref) => {
    return (
      <button
        ref={ref}
        {...props}
        disabled={isLoading || props.disabled}
        className="flex justify-center w-full px-4 py-2 text-sm font-medium text-white bg-indigo-600 border border-transparent rounded-md shadow-sm hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500 disabled:opacity-50"
      >
        {isLoading ? 'Loading...' : children}
      </button>
    );
  },
);

Button.displayName = 'Button';

export default Button;
