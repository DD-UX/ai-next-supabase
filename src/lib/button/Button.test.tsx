import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';

import Button from './Button';

// This is a comprehensive, type-safe mock for the 'motion/react' library.
// It is designed to pass a strict ESLint configuration.
jest.mock('motion/react', () => {
  // We must use require here as jest.mock is hoisted above imports.
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const React = require('react');

  // Define a type for the motion-specific props we want to filter out.
  interface MockMotionProps {
    children?: React.ReactNode;
    layout?: unknown;
    animate?: unknown;
    initial?: unknown;
    transition?: unknown;
  }

  // A generic factory for creating mocked motion components (e.g., motion.div).
  const createMock = (Tag: keyof JSX.IntrinsicElements) => {
    const Mock = React.forwardRef<HTMLElement, MockMotionProps>((props, ref) => {
      // Rename the motion props with a leading underscore to satisfy the 'no-unused-vars' rule.
      const {
        children,
        layout: _layout,
        animate: _animate,
        initial: _initial,
        transition: _transition,
        ...rest
      } = props;
      return React.createElement(Tag, { ...rest, ref }, children);
    });

    Mock.displayName = `MockMotion(${Tag})`;
    return Mock;
  };

  return {
    // We don't use requireActual to avoid pulling in 'any' types.
    // We explicitly mock everything we need.
    motion: {
      button: createMock('button'),
      div: createMock('div'),
      span: createMock('span'),
      svg: createMock('svg'),
    },
    useAnimate: () => [
      React.createRef<HTMLButtonElement>(),
      jest.fn().mockResolvedValue(undefined),
    ],
    AnimatePresence: ({ children }: { children: React.ReactNode }) =>
      React.createElement(React.Fragment, null, children),
  };
});

describe('Button Component', () => {
  it('renders with correct text', () => {
    render(<Button>Click Me</Button>);
    expect(screen.getByText('Click Me')).toBeInTheDocument();
  });

  it('applies the correct classes for each variant via snapshots', () => {
    const { rerender, asFragment } = render(
      <Button variant="primary">Primary</Button>,
    );
    expect(asFragment()).toMatchSnapshot('primary');

    rerender(<Button variant="secondary">Secondary</Button>);
    expect(asFragment()).toMatchSnapshot('secondary');

    rerender(<Button variant="link">Link</Button>);
    expect(asFragment()).toMatchSnapshot('link');

    rerender(<Button variant="success">Success</Button>);
    expect(asFragment()).toMatchSnapshot('success');

    rerender(<Button variant="warning">Warning</Button>);
    expect(asFragment()).toMatchSnapshot('warning');

    rerender(<Button variant="error">Error</Button>);
    expect(asFragment()).toMatchSnapshot('error');
  });

  it('calls onClick handler when clicked', () => {
    const mockOnClick = jest.fn();
    render(<Button onClick={mockOnClick}>Click Me</Button>);
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('is disabled when the disabled prop is true', () => {
    const mockOnClick = jest.fn();
    render(
      <Button onClick={mockOnClick} disabled>
        Disabled
      </Button>,
    );
    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
    fireEvent.click(button);
    expect(mockOnClick).not.toHaveBeenCalled();
  });

  it('shows loading and success states on async click and re-enables', async () => {
    jest.useFakeTimers();
    const mockOnClick = jest.fn().mockResolvedValue(undefined);
    render(<Button onClick={mockOnClick}>Submit</Button>);
    const button = screen.getByRole('button');

    expect(button).not.toBeDisabled();
    fireEvent.click(button);

    await waitFor(() => {
      expect(button).toBeDisabled();
    });

    await waitFor(() => {
      expect(screen.getByText('Loading, please wait.')).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(
        screen.getByText('Action completed successfully.'),
      ).toBeInTheDocument();
    });

    expect(button).toBeDisabled();

    await act(async () => {
      jest.advanceTimersByTime(2200);
    });

    expect(button).not.toBeDisabled();

    jest.useRealTimers();
  });

  it('does not show loading/success states if onClick is not provided', () => {
    render(<Button>No Action</Button>);
    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(button).not.toBeDisabled();

    expect(
      screen.queryByText('Loading, please wait.'),
    ).not.toBeInTheDocument();
    expect(
      screen.queryByText('Action completed successfully.'),
    ).not.toBeInTheDocument();
  });
});
