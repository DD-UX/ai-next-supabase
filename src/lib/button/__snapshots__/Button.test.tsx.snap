// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button Component applies the correct classes for each variant via snapshots: error 1`] = `
<DocumentFragment>
  <button
    class="flex min-w-[120px] cursor-pointer items-center justify-center gap-2 rounded-md px-4 py-2 text-sm font-medium ring-offset-2 transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed bg-red-600 text-white hover:bg-red-700 focus-visible:ring-red-500"
  >
    <div
      class="flex items-center gap-2"
    >
      <svg
        class="loader"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 3a9 9 0 1 0 9 9"
        />
      </svg>
      <svg
        class="check"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
        />
        <path
          d="M9 12l2 2l4 -4"
        />
      </svg>
      <span>
        Error
      </span>
      <span
        aria-atomic="true"
        aria-live="polite"
        class="sr-only"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button Component applies the correct classes for each variant via snapshots: link 1`] = `
<DocumentFragment>
  <button
    class="flex min-w-[120px] cursor-pointer items-center justify-center gap-2 rounded-md px-4 py-2 text-sm font-medium ring-offset-2 transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed text-indigo-600 underline-offset-4 hover:underline focus-visible:ring-indigo-500"
  >
    <div
      class="flex items-center gap-2"
    >
      <svg
        class="loader"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 3a9 9 0 1 0 9 9"
        />
      </svg>
      <svg
        class="check"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
        />
        <path
          d="M9 12l2 2l4 -4"
        />
      </svg>
      <span>
        Link
      </span>
      <span
        aria-atomic="true"
        aria-live="polite"
        class="sr-only"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button Component applies the correct classes for each variant via snapshots: primary 1`] = `
<DocumentFragment>
  <button
    class="flex min-w-[120px] cursor-pointer items-center justify-center gap-2 rounded-md px-4 py-2 text-sm font-medium ring-offset-2 transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed bg-indigo-600 text-white hover:bg-indigo-700 focus-visible:ring-indigo-500"
  >
    <div
      class="flex items-center gap-2"
    >
      <svg
        class="loader"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 3a9 9 0 1 0 9 9"
        />
      </svg>
      <svg
        class="check"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
        />
        <path
          d="M9 12l2 2l4 -4"
        />
      </svg>
      <span>
        Primary
      </span>
      <span
        aria-atomic="true"
        aria-live="polite"
        class="sr-only"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button Component applies the correct classes for each variant via snapshots: secondary 1`] = `
<DocumentFragment>
  <button
    class="flex min-w-[120px] cursor-pointer items-center justify-center gap-2 rounded-md px-4 py-2 text-sm font-medium ring-offset-2 transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed border border-slate-300 bg-transparent text-slate-900 hover:bg-slate-100 focus-visible:ring-slate-400"
  >
    <div
      class="flex items-center gap-2"
    >
      <svg
        class="loader"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 3a9 9 0 1 0 9 9"
        />
      </svg>
      <svg
        class="check"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
        />
        <path
          d="M9 12l2 2l4 -4"
        />
      </svg>
      <span>
        Secondary
      </span>
      <span
        aria-atomic="true"
        aria-live="polite"
        class="sr-only"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button Component applies the correct classes for each variant via snapshots: success 1`] = `
<DocumentFragment>
  <button
    class="flex min-w-[120px] cursor-pointer items-center justify-center gap-2 rounded-md px-4 py-2 text-sm font-medium ring-offset-2 transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed bg-green-600 text-white hover:bg-green-700 focus-visible:ring-green-500"
  >
    <div
      class="flex items-center gap-2"
    >
      <svg
        class="loader"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 3a9 9 0 1 0 9 9"
        />
      </svg>
      <svg
        class="check"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
        />
        <path
          d="M9 12l2 2l4 -4"
        />
      </svg>
      <span>
        Success
      </span>
      <span
        aria-atomic="true"
        aria-live="polite"
        class="sr-only"
      />
    </div>
  </button>
</DocumentFragment>
`;

exports[`Button Component applies the correct classes for each variant via snapshots: warning 1`] = `
<DocumentFragment>
  <button
    class="flex min-w-[120px] cursor-pointer items-center justify-center gap-2 rounded-md px-4 py-2 text-sm font-medium ring-offset-2 transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed bg-amber-500 text-white hover:bg-amber-600 focus-visible:ring-amber-500"
  >
    <div
      class="flex items-center gap-2"
    >
      <svg
        class="loader"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 3a9 9 0 1 0 9 9"
        />
      </svg>
      <svg
        class="check"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
      >
        <path
          d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
        />
        <path
          d="M9 12l2 2l4 -4"
        />
      </svg>
      <span>
        Warning
      </span>
      <span
        aria-atomic="true"
        aria-live="polite"
        class="sr-only"
      />
    </div>
  </button>
</DocumentFragment>
`;
